src/Bicep.Core/Workspaces/BicepFile.cs

```diff
@@ -3,19 +3,29 @@
 
 using System;
 using System.Collections.Immutable;
+using System.Linq;
+using Bicep.Core.Extensions;
 using Bicep.Core.Syntax;
+using Bicep.Core.Text;
 
 namespace Bicep.Core.Workspaces
 {
     public class BicepFile : ISourceFile
     {
+        private readonly Lazy<ImmutableDictionary<int, (int, ImmutableArray<string>)>> disableNextLineDiagnosticDirectivesCacheLazy;
+
         public BicepFile(Uri fileUri, ImmutableArray<int> lineStarts, ProgramSyntax programSyntax)
         {
             FileUri = fileUri;
             LineStarts = lineStarts;
             ProgramSyntax = programSyntax;
             Hierarchy = new SyntaxHierarchy();
             Hierarchy.AddRoot(ProgramSyntax);
+
+            var visitor = new SyntaxTriviaVisitor(LineStarts);
+            visitor.Visit(ProgramSyntax);
+
+            disableNextLineDiagnosticDirectivesCacheLazy = new Lazy<ImmutableDictionary<int, (int, ImmutableArray<string>)>>(() => visitor.GetDisableNextLineDiagnosticDirectivesCache());
```

Comment:
It would be good to have even the syntax visiting logic handled lazily so that it's done at most once for a single syntax tree. My suggestions:
* Put all of this logic into a separate class (`DisabledDiagnosticsCache`?), and add it as a property to `BicepFile` (similar to what's done with `SyntaxHierarchy`).
* Encapsulate all of this logic (creating the visitor, visiting the syntax, returning the dictionary) in a single `Lazy<T>`, inside the `DisabledDiagnosticsCache` class.
* In the constructor on line 31, simply copy the previous cache from the `original` bicep file - so you don't have to reconstruct it unnecessarily (similar to what is happening for the `Hierarchy` property)
* It would also be good to create a `record` for the dictionary value, rather than an anonymous tuple.

Reply:
Sure, updated as part of this commit:
https://github.com/Azure/bicep/commit/099a6ab1e68f807da4ec1510846a8b0b3b78655b